File: Inventor-Training-Material-0c232f96f36e1e0bed31b07a6a6ebb901a31dd40/Module 17 - Attributes/Attribute-VBNet/InvExe/Form1.vb
Imports Inventor
Imports System.Runtime.InteropServices
Imports System.Reflection
Imports System.Diagnostics


Public Class Form1

    Dim _InvApp As Inventor.Application
    Dim _macros As Macros

    Private Sub Button1_Click(sender As System.Object, e As System.EventArgs) Handles Button1.Click
        Try

            Dim memberName As String = ComboBoxMacros.SelectedItem.ToString()

            Dim params() As Object = Nothing
            _macros.GetType().InvokeMember(memberName, BindingFlags.InvokeMethod, Nothing, _macros, params, Nothing, Nothing, Nothing)

        Catch ex As Exception

            Dim Caption As String = ex.Message
            Dim Buttons As MessageBoxButtons = MessageBoxButtons.OK
            Dim Result As DialogResult = MessageBox.Show(ex.StackTrace, Caption, Buttons, MessageBoxIcon.Exclamation)

        End Try
    End Sub

    Private Sub Form1_Load(sender As System.Object, e As System.EventArgs) Handles MyBase.Load
        Try
            _InvApp = System.Runtime.InteropServices.Marshal.GetActiveObject("Inventor.Application")
        Catch ex As Exception
            MessageBox.Show("please open Inventor!")
            Exit Sub
        End Try


        ' Add any initialization after the InitializeComponent() call.
        _macros = New Macros(_InvApp)

        Dim methods As MemberInfo() = _macros.GetType().GetMembers()

        For Each member As MemberInfo In methods
            If (member.DeclaringType.Name = "Macros" And member.MemberType = MemberTypes.Method) Then
                ComboBoxMacros.Items.Add(member.Name)
            End If
        Next

        If ComboBoxMacros.Items.Count > 0 Then
            ComboBoxMacros.SelectedIndex = 0
            Button1.Enabled = True
        End If
    End Sub
End Class


'''macro class
Public Class Macros

    Dim _InvApplication As Inventor.Application

    Public Sub New(ByVal oApp As Inventor.Application)

        _InvApplication = oApp

    End Sub


    'Small helper function that prompts user for a file selection
    Private Function OpenFile(ByVal StrFilter As String) As String

        Dim filename As String = ""

        Dim ofDlg As System.Windows.Forms.OpenFileDialog = New System.Windows.Forms.OpenFileDialog()

        Dim user As String = System.Windows.Forms.SystemInformation.UserName

        ofDlg.Title = "Open File"
        ofDlg.InitialDirectory = "C:\Documents and Settings\" + user + "\Desktop\"

        ofDlg.Filter = StrFilter 'Example: "Inventor files (*.ipt; *.iam; *.idw)|*.ipt;*.iam;*.idw"
        ofDlg.FilterIndex = 1
        ofDlg.RestoreDirectory = True

        If (ofDlg.ShowDialog() = DialogResult.OK) Then
            filename = ofDlg.FileName
        End If

        OpenFile = filename

    End Function

    ''' <summary>
    ''' create an attribute
    ''' this assumes a part document with some features is opened.
    ''' </summary>
    ''' <remarks></remarks>
    Public Sub AddAttribute()

        ' Get the selected edge.  This assumes an edge is already selected.
        Dim oEdge As Edge
        oEdge = _InvApplication.ActiveDocument.SelectSet.Item(1)

        ' Add an attribute set to the edge.  If you only need to "name" the
        ' edge this is enough to find it later and you can skip the next step.
        Dim oAttribSet As AttributeSet
        oAttribSet = oEdge.AttributeSets.Add("BoltEdge")

        ' Add an attribute to the set.  This can be any information you
        ' want to associate with the edge.
        Call oAttribSet.Add("BoltRadius", ValueTypeEnum.kDoubleType, 0.5)

    End Sub

    ''' <summary>
    ''' Query Attribute
    '''  this assumes a part document with some features is opened.
    ''' and you have run  AddAttribute on an edge
    ''' </summary>
    ''' <remarks></remarks>
    Public Sub QueryAttribute()

        ' Set a reference to the attribute manager of the active document.
        Dim oAttribMgr As AttributeManager
        oAttribMgr = _InvApplication.ActiveDocument.AttributeManager

        ' Get the objects with a particular attribute attached.
        Dim oObjs As ObjectCollection
        oObjs = oAttribMgr.FindObjects("BoltEdge", "BoltRadius", 0.5)

        ' Get the objects that have an attribute set of a certain name.
        oObjs = oAttribMgr.FindObjects("BoltEdge")

        ' Get the attribute sets with a certain name.
        Dim oAttribSets As AttributeSetsEnumerator
        oAttribSets = oAttribMgr.FindAttributeSets("BoltEdge")

        ' Get the attribute sets with a certain name using a wild card.
        oAttribSets = oAttribMgr.FindAttributeSets("Bolt*")

    End Sub

    ''' <summary>
    ''' assumes a document is opened    ''' 
    ''' </summary>
    ''' <remarks></remarks>
    Public Sub Add_Persistent_Transient_Attribute()

        Dim oDoc As Document
        oDoc = _InvApplication.ActiveDocument


        MsgBox("add one attribute set named PersistentAttSet which is persistent" & vbCrLf & "add one attribute set named TransientAttSet which is transient")

            ' add persistent attribute
            Dim oPersistentAttSet As AttributeSet
            oPersistentAttSet = oDoc.AttributeSets.Add("PersistentAttSet")

            oPersistentAttSet.Add("PersistentAtt", ValueTypeEnum.kDoubleType, 0.5)
 

            'add transient attribute
            Dim oTransientAttSet As AttributeSet
            oTransientAttSet = oDoc.AttributeSets.AddTransient("TransientAttSet")

            oPersistentAttSet.Add("TransientAtt", ValueTypeEnum.kDoubleType, 1)


        MsgBox("now try to find the attributes PersistentAttSet in the current document")


        ' Set a reference to the attribute manager of the active document.
        Dim oAttribMgr As AttributeManager
        oAttribMgr = oDoc.AttributeManager

        Dim oAttSets As AttributeSetsEnumerator
        oAttSets = oAttribMgr.FindAttributeSets("PersistentAttSet")

        If oAttSets.Count > 0 Then
            MsgBox("find Persistent AttSet!")
        Else
            MsgBox("cannot find Persistent AttSet!")
        End If

        MsgBox("now try to find the attributes PersistentAttSet in the current document")

        oAttSets = oAttribMgr.FindAttributeSets("TransientAttSet")

        If oAttSets.Count > 0 Then
            MsgBox("find Transient AttSet!")
        Else
            MsgBox("cannot find Transient AttSet!")
        End If

        MsgBox("Please save and close this document, open it again. Then click [OK] of this dialog")

        oDoc = _InvApplication.ActiveDocument

        oAttribMgr = oDoc.AttributeManager 

        oAttSets = oAttribMgr.FindAttributeSets("PersistentAttSet")

        If oAttSets.Count > 0 Then
            MsgBox("find Persistent AttSet!")
        Else
            MsgBox("cannot find Persistent AttSet!")
        End If
 
        oAttSets = oAttribMgr.FindAttributeSets("TransientAttSet")

        If oAttSets.Count > 0 Then
            MsgBox("find Transient AttSet!")
        Else
            MsgBox("cannot find Transient AttSet!")
        End If


    End Sub


End Class

File: Inventor-Training-Material-0c232f96f36e1e0bed31b07a6a6ebb901a31dd40/Module 17 - Attributes/Attributes_C#/InvEXE/InvEXE/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.Runtime.InteropServices;
using Inventor;
 
using System.Diagnostics;
using System.Reflection;

namespace InvEXE
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }
        private Inventor.Application _InvApplication;
        Macros _macros;
        private void Form1_Load(object sender, EventArgs e)
        {

            try
            {
                _InvApplication = (Inventor.Application)Marshal.GetActiveObject("Inventor.Application");
            }
            catch (Exception ex)
            {
                MessageBox.Show("make sure an Inventor instance is launched!");
                return;
            }

            // Add any initialization after the InitializeComponent() call.
            _macros = new Macros(_InvApplication);

            MemberInfo[] methods = _macros.GetType().GetMembers();

            foreach (MemberInfo member in methods)
            {
                if ((member.DeclaringType.Name == "Macros" & member.MemberType == MemberTypes.Method))
                {
                    ComboBoxMacros.Items.Add(member.Name);
                }
            }

            if (ComboBoxMacros.Items.Count > 0)
            {
                ComboBoxMacros.SelectedIndex = 0;
                button1.Enabled = true;
            }

 
        }

        private void button1_Click(object sender, EventArgs e)
        {
            try
            {

                string memberName = ComboBoxMacros.SelectedItem.ToString();

                object[] @params = null;
                _macros.GetType().InvokeMember(memberName, BindingFlags.InvokeMethod, null, _macros, @params, null, null, null);
            }

            catch (Exception ex)
            {

                string Caption = ex.Message;
                MessageBoxButtons Buttons = MessageBoxButtons.OK;
                DialogResult Result = MessageBox.Show(ex.StackTrace, Caption, Buttons, MessageBoxIcon.Exclamation);

            } 
        }
    }

       //macro class
    public class Macros
    {

        Inventor.Application _InvApplication;

        public Macros(Inventor.Application oApp)
        {
            _InvApplication = oApp;
        }

        //Small helper function that prompts user for a file selection
        private string OpenFile(string StrFilter)
        {

            string filename = "";

            System.Windows.Forms.OpenFileDialog ofDlg = new System.Windows.Forms.OpenFileDialog();

            string user = System.Windows.Forms.SystemInformation.UserName;

            ofDlg.Title = "Open File";
            ofDlg.InitialDirectory = "C:\\Documents and Settings\\" + user + "\\Desktop\\";

            ofDlg.Filter = StrFilter;
            //Example: "Inventor files (*.ipt; *.iam; *.idw)|*.ipt;*.iam;*.idw"
            ofDlg.FilterIndex = 1;
            ofDlg.RestoreDirectory = true;

            if ((ofDlg.ShowDialog() == DialogResult.OK))
            {
                filename = ofDlg.FileName;
            }

            return filename;

        }

        /// <summary>
    /// create an attribute
    /// this assumes a part document with some features is opened.
    /// </summary>
    /// <remarks></remarks>

    public void AddAttribute()
    {
     // Get the selected edge.  This assumes an edge is already selected.
     Edge oEdge  = (Edge)_InvApplication.ActiveDocument.SelectSet[1];

     // Add an attribute set to the edge.  If you only need to "name" the
     // edge this is enough to find it later and you can skip the next step.
     AttributeSet oAttribSet = oEdge.AttributeSets.Add("BoltEdge");

     // Add an attribute to the set.  This can be any information you
     // want to associate with the edge.
     oAttribSet.Add("BoltRadius", ValueTypeEnum.kDoubleType, 0.5);

    }

        /// <summary>
        /// Query Attribute
        ///  this assumes a part document with some features is opened.
        /// and you have run  AddAttribute on an edge
        /// </summary>
        /// <remarks></remarks>

        public void QueryAttribute()
        {
         // Set a reference to the attribute manager of the active document.
         AttributeManager oAttribMgr = _InvApplication.ActiveDocument.AttributeManager;

         // Get the objects with a particular attribute attached.
         ObjectCollection oObjs  = oAttribMgr.FindObjects("BoltEdge", "BoltRadius", 0.5);

         // Get the objects that have an attribute set of a certain name.
         oObjs = oAttribMgr.FindObjects("BoltEdge");

         // Get the attribute sets with a certain name.
         AttributeSetsEnumerator oAttribSets   = oAttribMgr.FindAttributeSets("BoltEdge");

         // Get the attribute sets with a certain name using a wild card.
         oAttribSets = oAttribMgr.FindAttributeSets("Bolt*");

        }

        /// <summary>
        /// assumes a document is opened    ''' 
        /// </summary>
        /// <remarks></remarks>

        public void Add_Persistent_Transient_Attribute()
        {
            Document oDoc = _InvApplication.ActiveDocument;


            MessageBox.Show("add one attribute set named PersistentAttSet which is persistent" + "\n" + "add one attribute set named TransientAttSet which is transient");

            // add persistent attribute
            AttributeSet oPersistentAttSet = oDoc.AttributeSets.Add("PersistentAttSet");

            oPersistentAttSet.Add("PersistentAtt", ValueTypeEnum.kDoubleType, 0.5);


            //add transient attribute
            AttributeSet oTransientAttSet = oDoc.AttributeSets.AddTransient("TransientAttSet");

            oPersistentAttSet.Add("TransientAtt", ValueTypeEnum.kDoubleType, 1);


            MessageBox.Show("now try to find the attributes PersistentAttSet in the current document");


            // Set a reference to the attribute manager of the active document.
            AttributeManager oAttribMgr = oDoc.AttributeManager;

            AttributeSetsEnumerator oAttSets = oAttribMgr.FindAttributeSets("PersistentAttSet");

            if (oAttSets.Count > 0)
            {
                MessageBox.Show("find Persistent AttSet!");
            }
            else
            {
                MessageBox.Show("cannot find Persistent AttSet!");
            }

            MessageBox.Show("now try to find the attributes PersistentAttSet in the current document");

            oAttSets = oAttribMgr.FindAttributeSets("TransientAttSet");

            if (oAttSets.Count > 0)
            {
                MessageBox.Show("find Transient AttSet!");
            }
            else
            {
                MessageBox.Show("cannot find Transient AttSet!");
            }

            MessageBox.Show("Please save and close this document, open it again. Then click [OK] of this dialog");

            oDoc = _InvApplication.ActiveDocument;

            oAttribMgr = oDoc.AttributeManager;

            oAttSets = oAttribMgr.FindAttributeSets("PersistentAttSet");

            if (oAttSets.Count > 0)
            {
                MessageBox.Show("find Persistent AttSet!");
            }
            else
            {
                MessageBox.Show("cannot find Persistent AttSet!");
            }

            oAttSets = oAttribMgr.FindAttributeSets("TransientAttSet");

            if (oAttSets.Count > 0)
            {
                MessageBox.Show("find Transient AttSet!");
            }
            else
            {
                MessageBox.Show("cannot find Transient AttSet!");
            }
        }

    }
}
